File: task_1/cpp/1(3).cpp
#include <iostream>
#include <cmath>
#include <iomanip>

using namespace std;

// Исходная функция
double f(double x) {
    return x * log(x + 1) - 1;
}

// Первая производная
double df(double x) {
    return log(x + 1) + x / (x + 1);
}

// Вторая производная
double ddf(double x) {
    return 1.0 / (x + 1) + 1.0 / ((x + 1) * (x + 1));
}

// Метод Ньютона
double NewtonMethod(double x, double eps) {
    int k = 0;
    double prevX = x - 2 * eps;

    // Проверка условия сходимости
    if (f(x) * ddf(x) <= 0) {
        cout << "Условие сходимости нарушено при x0 = " << x << ". Попробуйте другое приближение.\n";
        return NAN;
    }

    cout << "╔═════════════════════════════════════════════════════════════════════════════╗\n";
    cout << "║   N   ║         x_n           ║         x_(n+1)       ║    x_(n+1) - x_n    ║\n";
    cout << "╠═════════════════════════════════════════════════════════════════════════════╣\n";

    while (fabs(x - prevX) > eps) {
        prevX = x;
        x = x - f(x) / df(x);
        cout << fixed << setprecision(6)
        << "║  " << setw(3) << k << "  ║\t"
        << prevX << "\t║\t" << x << "\t║\t"
        << fabs(x - prevX) << "      ║" << "\n";
        ++k;
    }

    cout << "╚═════════════════════════════════════════════════════════════════════════════╝\n";
    return x;
}

int main() {
    double eps = 1e-4;

    cout << "\n=== Поиск положительного корня ===\n";
    double x0_pos = 1.5;
    double root_pos = NewtonMethod(x0_pos, eps);
    if (!isnan(root_pos))
        cout << "\nПоложительный корень (Ньютон): x ≈ " << setprecision(6) << root_pos << "\n";

    cout << "\n=== Поиск отрицательного корня ===\n";
    double x0_neg = -0.8;
    double root_neg = NewtonMethod(x0_neg, eps);
    if (!isnan(root_neg))
        cout << "\nОтрицательный корень (Ньютон): x ≈ " << setprecision(6) << root_neg << "\n";

    return 0;
}

File: task_1/go/1(3).go
package main

import (
	"fmt"
	"math"
)

// Исходная функция
func f(x float64) float64 {
	return x * math.Log(x+1) - 1
}

// Первая производная
func df(x float64) float64 {
	return math.Log(x+1) + x/(x+1)
}

// Вторая производная
func ddf(x float64) float64 {
	return 1.0/(x+1) + 1.0/((x+1)*(x+1))
}

// Метод Ньютона
func NewtonMethod(x, eps float64) float64 {
	k := 0
	prevX := x - 2*eps

	// Проверка условия сходимости
	if f(x)*ddf(x) <= 0 {
		fmt.Printf("Условие сходимости нарушено при x0 = %.6f. Попробуйте другое приближение.\n", x)
		return math.NaN()
	}

	fmt.Println("╔═════════════════════════════════════════════════════════════════════════════╗")
	fmt.Println("║   N   ║         x_n           ║         x_(n+1)       ║    x_(n+1) - x_n    ║")
	fmt.Println("╠═════════════════════════════════════════════════════════════════════════════╣")

	for math.Abs(x-prevX) > eps {
		prevX = x
		x = x - f(x)/df(x)
		fmt.Printf("║  %3d  ║\t%.6f\t║\t%.6f\t║\t%.6f      ║\n", k, prevX, x, math.Abs(x-prevX))
		k++
	}

	fmt.Println("╚═════════════════════════════════════════════════════════════════════════════╝")
	return x
}

func main() {
	eps := 1e-4

	fmt.Println("\n=== Поиск положительного корня ===")
	x0Pos := 1.5
	rootPos := NewtonMethod(x0Pos, eps)
	if !math.IsNaN(rootPos) {
		fmt.Printf("\nПоложительный корень (Ньютон): x ≈ %.6f\n", rootPos)
	}

	fmt.Println("\n=== Поиск отрицательного корня ===")
	x0Neg := -0.8
	rootNeg := NewtonMethod(x0Neg, eps)
	if !math.IsNaN(rootNeg) {
		fmt.Printf("\nОтрицательный корень (Ньютон): x ≈ %.6f\n", rootNeg)
	}
}
